### PR TITLE
ci: use .java-version and prepare to use VM with installed tools in BK

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

* Enable .java-version - that contains the current major java version
* Enable `dry-run` in the release/snapshot by default.
* Fallback to install java in Buildkite if the JAVA_HOME is not in the Buildkite runner - a follow-up to use a VM with the installed tools

### Why

* Stability and reliability.
* Faster builds.

### Follow-up

- Use the VMs with the tools in Buildkite

### Test

Triggering manually against this particular feature branch (for such, I use the snapshot one):
- Snapshot [build](https://buildkite.com/elastic/ecs-logging-java-snapshot/builds/36) in Buildkite
- Release [build](https://buildkite.com/elastic/ecs-logging-java-release/builds/8) in Buildkite

### CODE REVIEW

File: .ci/snapshot.sh [hunk 1]
```diff
@@ -24,9 +24,14 @@ echo $PATH
...
+GOAL="install"
+DRY_RUN_MSG="(dry-run)"
+# Otherwise, a RELEASE
+if [[ "$dry_run" == "false" ]] ; then
+  GOAL="deploy"
+  DRY_RUN_MSG=""
 fi
+
+echo "--- Deploy the snapshot :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
+./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee release.txt
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:37]:
```suggestion
./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee snapshot.txt
```

File: .ci/snapshot.sh [hunk 2]
```diff
@@ -24,9 +24,14 @@ echo $PATH
 java -version
 
 set +x
-echo "--- Deploy the snapshot :package:"
-if [[ "$dry_run" == "true" ]] ; then
-  ./mvnw -V -s .ci/settings.xml -Pgpg clean package --batch-mode | tee snapshot.txt
-else
-  ./mvnw -V -s .ci/settings.xml -Pgpg clean deploy --batch-mode | tee snapshot.txt
+# Default in dry-run mode
+GOAL="install"
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:28]: For dry-run, `package` is enough

```suggestion
GOAL="package"
```

File: .ci/release.sh [hunk 3]
```diff
@@ -27,9 +27,14 @@ echo $PATH
 java -version
 
 set +x
-echo "--- Release the binaries to Maven Central :maven:"
-if [[ "${dry_run}" == "true" ]] ; then
-  ./mvnw -V -s .ci/settings.xml -Pgpg clean package --batch-mode | tee release.txt
-else
-  ./mvnw -V -s .ci/settings.xml -Pgpg clean deploy --batch-mode | tee release.txt
+# Default in dry-run mode
+GOAL="install"
```
Comment (SylvainJuge) [inline .ci/release.sh:31]: For dry-run, `package` is enough

```suggestion
GOAL="package"
```

**Review by amannocci — APPROVED** [review]

**Review by amannocci — APPROVED** [review]

File: .buildkite/hooks/pre-command [hunk 4]
```diff
@@ -45,16 +45,23 @@ echo "--- Configure git context :git:"
...
-curl -L --output /tmp/jdk.tar.gz "$JAVA_PKG"
-mkdir -p "$JAVA_HOME"
-tar --extract --file /tmp/jdk.tar.gz --directory "$JAVA_HOME" --strip-components 1
-
+# Configure the java version
+JAVA_VERSION=$(cat .java-version | xargs | tr -dc '[:print:]')
+JAVA_HOME="${HOME}/.java/openjdk${JAVA_VERSION}"
 export JAVA_HOME
-PATH=$JAVA_HOME/bin:$PATH
+PATH="${JAVA_HOME}/bin:$PATH"
```
Comment (SylvainJuge) [inline .buildkite/hooks/pre-command:52]:
```suggestion
PATH="${JAVA_HOME}/bin:${PATH}"
```

File: .buildkite/hooks/pre-command [hunk 5]
```diff
@@ -45,16 +45,23 @@ echo "--- Configure git context :git:"
...
 
-java -version || true
+# Fallback to install at runtime
+if [ ! -d "${JAVA_HOME}" ] ; then
+  # This should not be the case normally untless the .java-version file has been changed
+  # and the VM Image is not yet available with the latest version.
+  echo "--- Install JDK${JAVA_VERSION} :java:"
+  JAVA_URL=https://jvm-catalog.elastic.co/jdk
+  JAVA_PKG="${JAVA_URL}/latest_openjdk_${JAVA_VERSION}_linux.tar.gz"
+  curl -L --output /tmp/jdk.tar.gz "$JAVA_PKG"
```
Comment (SylvainJuge) [inline .buildkite/hooks/pre-command:62]:
```suggestion
  curl -L --output /tmp/jdk.tar.gz "${JAVA_PKG}"
```

File: .buildkite/hooks/pre-command [hunk 6]
```diff
@@ -45,16 +45,23 @@ echo "--- Configure git context :git:"
...
+  mkdir -p "$JAVA_HOME"
+  tar --extract --file /tmp/jdk.tar.gz --directory "$JAVA_HOME" --strip-components 1
```
Comment (SylvainJuge) [inline .buildkite/hooks/pre-command:64]:
```suggestion
  tar --extract --file /tmp/jdk.tar.gz --directory "${JAVA_HOME}" --strip-components 1
```

File: .ci/release.sh [hunk 7]
```diff
@@ -27,9 +27,14 @@ echo $PATH
...
+# Default in dry-run mode
+GOAL="package"
+DRY_RUN_MSG="(dry-run)"
+# Otherwise, a RELEASE
+if [[ "$dry_run" == "false" ]] ; then
+  GOAL="deploy"
+  DRY_RUN_MSG=""
 fi
+
+echo "--- Release the binaries to Maven Central :maven: [./mvnw $GOAL)] $DRY_RUN_MSG"
```
Comment (SylvainJuge) [inline .ci/release.sh:39]:
```suggestion
echo "--- Release the binaries to Maven Central :maven: [./mvnw ${GOAL})] ${DRY_RUN_MSG}"
```

File: .ci/release.sh [hunk 8]
```diff
@@ -27,9 +27,14 @@ echo $PATH
...
+./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee release.txt
```
Comment (SylvainJuge) [inline .ci/release.sh:40]:
```suggestion
./mvnw -V -s .ci/settings.xml -Pgpg clean ${GOAL} -DskipTests --batch-mode | tee release.txt
```

File: .ci/snapshot.sh [hunk 9]
```diff
@@ -24,9 +24,14 @@ echo $PATH
...
+GOAL="package"
+DRY_RUN_MSG="(dry-run)"
+# Otherwise, a RELEASE
+if [[ "$dry_run" == "false" ]] ; then
+  GOAL="deploy"
+  DRY_RUN_MSG=""
 fi
+
+echo "--- Deploy the snapshot :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
+./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee snapshot.txt
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:37]:
```suggestion
echo "--- Deploy the snapshot :package: [./mvnw ${GOAL})] ${DRY_RUN_MSG}"
./mvnw -V -s .ci/settings.xml -Pgpg clean ${GOAL} -DskipTests --batch-mode | tee snapshot.txt
```

**Review by SylvainJuge — DISMISSED** [review]

**Review by reakaleek — DISMISSED** [review]

**Review by SylvainJuge — APPROVED** [review]

**Comment by v1v** [comment]
run docs build

**Comment by v1v** [comment]
run docs-build rebuild

**Comment by v1v** [comment]
run docs-build rebuild

**Comment by v1v** [comment]
run docs-build

**Pull request merged**